### PR TITLE
WebSocket 2.1: Implement upgradeHttpToWebSocket and return full URI from getRequestURI

### LINKED ISSUE #20584: WebSocket 2.1: Remove the restriction that, in a Jakarta web container environment, endpoints can only registered during the deployment of the web application. Also add a new method, ServerContainer.upgradeHttpToWebSocket() that allows a web application to programmatically dispatch a request to a WebSocket endpoint.
Spec issue: `https://github.com/jakartaee/websocket/issues/211`

See "A.1. Changes Between 2.1 and 2.0" here: https://github.com/jakartaee/websocket/blob/master/spec/src/main/asciidoc/WebSocket.adoc

### LINKED ISSUE #20585: WebSocket 2.1: Clarify the expected behaviour for Session.getRequestURI(). The full URI should be returned
Spec issue: `https://github.com/jakartaee/websocket/issues/228`

See "A.1. Changes Between 2.1 and 2.0" here: https://github.com/jakartaee/websocket/blob/master/spec/src/main/asciidoc/WebSocket.adoc

### PR DESCRIPTION
**Pull request by volosied**
fixes #20584 

1) I removed restrictions to when endpoints could be added (checks look for websocket 2.1)  (Tests required -- issue #22364) 

2) Created a new ServletContainerFactory class for the 1.0, 1.1, 2.0  features (default impl).  And another factory for the 2.1 to handle the upgradeHttpToWebSocket method. 

fixes #20585

1) Build the full URI via req.getRequestURL().toString();

### CODE REVIEW

**Comment by volosied** [comment]
#build

File: dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/HandshakeProcessor.java [hunk 1]
```diff
@@ -475,4 +479,33 @@ public static List<Extension> parseClientExtensions(List<String> list) {
...
+            String http = "http";
+
+            if(url.startsWith(https)){
+                url = "wss" + url.substring(https.length(), url.length());
+            } else if(url.startsWith(http)){
+                url = "ws" + url.substring(http.length(), url.length());
+            }
+
+            if( !(url.startsWith("ws:") || url.startsWith("wss:")) ){
+                throw new Exception("Scheme is not of type ws or wss.");
```
Comment (volosied) [inline dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/HandshakeProcessor.java]: @pnicolucci  Let me know if I should do something else with this exception.

Comment (volosied) [inline dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/HandshakeProcessor.java]: Add a message and request translation

File: dev/io.openliberty.wsoc.2.1.internal/src/io/openliberty/wsoc/servercontainer21/ServerContainerExt21.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        }
+
+        if(!(httpServletResponse instanceof HttpServletResponse)){
+            throw new DeploymentException("httpServletResponse not of type HttpServletResponse");
+        }
+
+        try {
+            doUpgrade((HttpServletRequest) httpServletRequest,(HttpServletResponse) httpServletResponse, sec, pathParameters);
+        } catch (ServletException ex){
+          throw new DeploymentException(ex.getCause().toString());
```
Comment (volosied) [inline dev/io.openliberty.wsoc.2.1.internal/src/io/openliberty/wsoc/servercontainer21/ServerContainerExt21.java]: Message & Translation

**Comment by volosied** [comment]
#build

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=io.openliberty.wsoc.internal_fat,io.openliberty.wsoc.2.1.internal_fat

File: dev/com.ibm.ws.wsoc/resources/com/ibm/ws/wsoc/internal/resources/WebSockets.nlsprops [hunk 3]
```diff
@@ -229,3 +229,11 @@ missingslashornull.uri.useraction=Define non null Endpoint URI beginning with /.
...
 wsoc.feature.not.loaded.correctly.explanation=The websocket feature is loaded with a version that corresponds to the Java [Jakarta] EE specification of that websocket feature. The feature was loaded with no version defined.
 wsoc.feature.not.loaded.correctly.useraction=Check the server.xml file for the existence of multiple websocket features. Only one should be specified.
+
+scheme.incorrect.error=CWWKH0054E: An unidenified scheme has been used in the uri "{0}". 
+scheme.incorrect.error.explanation=An unidenified scheme was found during the handling of the request. 
+scheme.incorrect.error.useraction=The scheme must be of type http, https, ws, or wss. 
+
+upgradeHttpToWebSocket.failed=CWWKH0055E: The jakarta.websocket.server.ServerContainer#upgradeHttpToWebSocket call failed due to {0}. 
+upgradeHttpToWebSocket.failed.explanation=The server could not upgrade the http(s) request to the ws/wss protocol.
+upgradeHttpToWebSocket.failed.useraction=Review the error for the to identitfy why the request failed to upgrade.
```
Comment (pnicolucci) [inline dev/com.ibm.ws.wsoc/resources/com/ibm/ws/wsoc/internal/resources/WebSockets.nlsprops]: "for the to" doesn't make sense to me. Can we reword this? Has ID reviewed?

File: dev/io.openliberty.wsoc.2.1.internal_fat/test-applications/basic21.war/src/basic/war/UpgradeServlet.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
+
+/*******************************************************************************
+ * Copyright (c) 2014 IBM Corporation and others.
```
Comment (pnicolucci) [inline dev/io.openliberty.wsoc.2.1.internal_fat/test-applications/basic21.war/src/basic/war/UpgradeServlet.java]: Is this copy right correct?

File: dev/io.openliberty.wsoc.2.1.internal/bnd.bnd [hunk 5]
```diff
@@ -17,9 +17,27 @@ Bundle-Name: WebSocket
...
+ Private-Package: \
+  io.openliberty.wsoc.servercontainer21
+
+ -buildpath: \
+ 	com.ibm.ws.wsoc.jakarta;version=latest,\
+ 	io.openliberty.jakarta.websocket.client.2.1;version=latest,\
+ 	io.openliberty.jakarta.websocket.2.1;version=latest,\
+ 	io.openliberty.jakarta.servlet.6.0;version=latest,\
+ 	com.ibm.ws.webcontainer;version=latest, \
+  com.ibm.ws.logging;version=latest,\
```
Comment (pnicolucci) [inline dev/io.openliberty.wsoc.2.1.internal/bnd.bnd]: Can you clean up this formatting?

File: dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/servercontainer/v10/ServerContainerImplFactory10.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ *******************************************************************************/
+package com.ibm.ws.wsoc.servercontainer.v10;
+
+import com.ibm.ws.wsoc.servercontainer.ServerContainerExt;
+import com.ibm.ws.wsoc.servercontainer.ServletContainerFactory;
+
+/*
+ * Used for Websocket 1.0, 1.1 and 2.0
+ */
+public class ServerContainerImplFactory10 implements ServletContainerFactory {
```
Comment (pnicolucci) [inline dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/servercontainer/v10/ServerContainerImplFactory10.java:19]: I see we have the exact same class: dev/io.openliberty.wsoc.1.0.internal/src/io/openliberty/wsoc/servercontainer10/ServerContainerImplFactory10.java are both of these necessary?

File: dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/WebSocketServletContainerInitializer.java [hunk 7]
```diff
@@ -28,11 +28,12 @@
 import com.ibm.websphere.ras.Tr;
```
Comment (pnicolucci) [inline dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/WebSocketServletContainerInitializer.java:28]: Copyright update?

File: dev/com.ibm.ws.wsoc/resources/com/ibm/ws/wsoc/internal/resources/WebSockets.nlsprops [hunk 8]
```diff
@@ -229,3 +229,11 @@ missingslashornull.uri.useraction=Define non null Endpoint URI beginning with /.
 wsoc.feature.not.loaded.correctly=CWWKH0053E: The websocket feature did not load with an acceptable version.
```
Comment (pnicolucci) [inline dev/com.ibm.ws.wsoc/resources/com/ibm/ws/wsoc/internal/resources/WebSockets.nlsprops:229]: Copyright update as well.

**Comment by volosied** [comment]
#libby

File: dev/com.ibm.ws.wsoc/resources/com/ibm/ws/wsoc/internal/resources/WebSockets.nlsprops [hunk 9]
```diff
@@ -229,3 +229,11 @@ missingslashornull.uri.useraction=Define non null Endpoint URI beginning with /.
...
 wsoc.feature.not.loaded.correctly.explanation=The websocket feature is loaded with a version that corresponds to the Java [Jakarta] EE specification of that websocket feature. The feature was loaded with no version defined.
 wsoc.feature.not.loaded.correctly.useraction=Check the server.xml file for the existence of multiple websocket features. Only one should be specified.
+
+scheme.incorrect.error=CWWKH0054E: An unidenified scheme has been used in the uri "{0}".
```
Comment (dmuelle) [inline dev/com.ibm.ws.wsoc/resources/com/ibm/ws/wsoc/internal/resources/WebSockets.nlsprops]:
```suggestion
scheme.incorrect.error=CWWKH0054E: An unidentified scheme was used in the {0} URI. 
```

File: dev/com.ibm.ws.wsoc/resources/com/ibm/ws/wsoc/internal/resources/WebSockets.nlsprops [hunk 10]
```diff
@@ -229,3 +229,11 @@ missingslashornull.uri.useraction=Define non null Endpoint URI beginning with /.
 wsoc.feature.not.loaded.correctly=CWWKH0053E: The websocket feature did not load with an acceptable version.
 wsoc.feature.not.loaded.correctly.explanation=The websocket feature is loaded with a version that corresponds to the Java [Jakarta] EE specification of that websocket feature. The feature was loaded with no version defined.
 wsoc.feature.not.loaded.correctly.useraction=Check the server.xml file for the existence of multiple websocket features. Only one should be specified.
+
+scheme.incorrect.error=CWWKH0054E: An unidenified scheme has been used in the uri "{0}". 
+scheme.incorrect.error.explanation=An unidenified scheme was found during the handling of the request.
```
Comment (dmuelle) [inline dev/com.ibm.ws.wsoc/resources/com/ibm/ws/wsoc/internal/resources/WebSockets.nlsprops]:
```suggestion
scheme.incorrect.error.explanation=An unidentified scheme was found during the handling of the request. 
```

File: dev/com.ibm.ws.wsoc/resources/com/ibm/ws/wsoc/internal/resources/WebSockets.nlsprops [hunk 11]
```diff
@@ -229,3 +229,11 @@ missingslashornull.uri.useraction=Define non null Endpoint URI beginning with /.
 wsoc.feature.not.loaded.correctly=CWWKH0053E: The websocket feature did not load with an acceptable version.
 wsoc.feature.not.loaded.correctly.explanation=The websocket feature is loaded with a version that corresponds to the Java [Jakarta] EE specification of that websocket feature. The feature was loaded with no version defined.
 wsoc.feature.not.loaded.correctly.useraction=Check the server.xml file for the existence of multiple websocket features. Only one should be specified.
+
+scheme.incorrect.error=CWWKH0054E: An unidenified scheme has been used in the uri "{0}". 
+scheme.incorrect.error.explanation=An unidenified scheme was found during the handling of the request. 
+scheme.incorrect.error.useraction=The scheme must be of type http, https, ws, or wss.
```
Comment (dmuelle) [inline dev/com.ibm.ws.wsoc/resources/com/ibm/ws/wsoc/internal/resources/WebSockets.nlsprops]:
```suggestion
scheme.incorrect.error.useraction=The scheme type must be http, https, ws, or wss. 
```

**Review by dmuelle — CHANGES_REQUESTED** [review]
ID review, let me know if you have any questions, thanks.

**Comment by posmith** [comment]
L2 message review completed

File: dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/HandshakeProcessor.java (hunk 1 above)
Comment (pnicolucci) [reply dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/HandshakeProcessor.java]: If this is a user-visible exception it should be translated, we can do that in a follow on PR if you open up an issue so we can address it after we merge this one.

Comment (pnicolucci) [reply dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/HandshakeProcessor.java]: @volosied  looks like you have already created a message! So we can resolve this one.

File: dev/io.openliberty.wsoc.2.1.internal_fat/test-applications/basic21.war/src/basic/war/UpgradeServlet.java [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    @Override
+    public void doGet(HttpServletRequest req, HttpServletResponse resp)
+                    throws ServletException, IOException {
+
+        ServerContainer container = (ServerContainer) req.getServletContext().getAttribute("jakarta.websocket.server.ServerContainer");
+        try{
+            // echo endpoint is used into another test, so I am reusing it here. 
+            container.upgradeHttpToWebSocket(req, resp, ServerEndpointConfig.Builder.create(EchoServerEP.class, "/echo").build(), new HashMap<String, String>());
+        } catch(DeploymentException ex){
+
```
Comment (pnicolucci) [inline dev/io.openliberty.wsoc.2.1.internal_fat/test-applications/basic21.war/src/basic/war/UpgradeServlet.java:42]: Should we at least log the exception so we know what happened during an exception scenario during testing? I'm fine doing this in a follow-on issue if you open the issue now and w e address it soon.

File: dev/io.openliberty.wsoc.2.1.internal/src/io/openliberty/wsoc/servercontainer21/ServerContainerExt21.java (hunk 2 above)
Comment (pnicolucci) [reply dev/io.openliberty.wsoc.2.1.internal/src/io/openliberty/wsoc/servercontainer21/ServerContainerExt21.java]: Looks good.

File: dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/servercontainer/v10/ServerContainerImplFactory10.java (hunk 6 above)
Comment (volosied) [reply dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/servercontainer/v10/ServerContainerImplFactory10.java:19]: You're right. 

I thought I deleted the one under io.openliberty.wsoc.1.0.internal, but I hadn't. It's now removed in the last commit.

File: dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/HandshakeProcessor.java (hunk 1 above)
Comment (volosied) [reply dev/com.ibm.ws.wsoc/src/com/ibm/ws/wsoc/HandshakeProcessor.java]: Ye

File: dev/io.openliberty.wsoc.2.1.internal_fat/test-applications/basic21.war/src/basic/war/UpgradeServlet.java (hunk 12 above)
Comment (volosied) [reply dev/io.openliberty.wsoc.2.1.internal_fat/test-applications/basic21.war/src/basic/war/UpgradeServlet.java:42]: Let's do a follow up

Comment (volosied) [reply dev/io.openliberty.wsoc.2.1.internal_fat/test-applications/basic21.war/src/basic/war/UpgradeServlet.java:42]: #22404

File: dev/io.openliberty.wsoc.2.1.internal_fat/test-applications/basic21.war/src/basic/war/UpgradeServlet.java (hunk 4 above)
Comment (volosied) [reply dev/io.openliberty.wsoc.2.1.internal_fat/test-applications/basic21.war/src/basic/war/UpgradeServlet.java]: Oops. I updated WsocUpgradeServlet.java instead.  This is now properly fixed.

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=io.openliberty.wsoc.internal_fat,io.openliberty.wsoc.2.1.internal_fat

**Comment by pnicolucci** [comment]
@dmuelle changes have been made if you can take a final look and approve? Thanks!

**Review by pnicolucci — APPROVED** [review]

File: dev/com.ibm.ws.wsoc/resources/com/ibm/ws/wsoc/internal/resources/WebSockets.nlsprops [hunk 13]
```diff
@@ -229,3 +229,11 @@ missingslashornull.uri.useraction=Define non null Endpoint URI beginning with /.
 wsoc.feature.not.loaded.correctly=CWWKH0053E: The websocket feature did not load with an acceptable version.
 wsoc.feature.not.loaded.correctly.explanation=The websocket feature is loaded with a version that corresponds to the Java [Jakarta] EE specification of that websocket feature. The feature was loaded with no version defined.
 wsoc.feature.not.loaded.correctly.useraction=Check the server.xml file for the existence of multiple websocket features. Only one should be specified.
+
+scheme.incorrect.error=CWWKH0054E: An unidentified scheme was used in the {0} URI.
+scheme.incorrect.error.explanation=An unidentified scheme was found during the handling of the request. 
+scheme.incorrect.error.useraction=The scheme type must be http, https, ws, or wss. 
+
+upgradeHttpToWebSocket.failed=CWWKH0055E: The jakarta.websocket.server.ServerContainer#upgradeHttpToWebSocket call failed due to {0}. 
+upgradeHttpToWebSocket.failed.explanation=The server could not upgrade the http(s) request to the ws/wss protocol.
```
Comment (dmuelle) [inline dev/com.ibm.ws.wsoc/resources/com/ibm/ws/wsoc/internal/resources/WebSockets.nlsprops]:
```suggestion
upgradeHttpToWebSocket.failed.explanation=The server could not upgrade the HTTP or HTTPS request to the ws/wss protocol.
```

Comment (volosied) [reply dev/com.ibm.ws.wsoc/resources/com/ibm/ws/wsoc/internal/resources/WebSockets.nlsprops]: Updated, thanks!

**Review by dmuelle — APPROVED** [review]
LGTM, thanks

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=io.openliberty.wsoc.internal_fat,io.openliberty.wsoc.2.1.internal_fat

**Comment by volosied** [comment]
#build #spawn.fullfat.buckets=io.openliberty.wsoc.internal_fat,io.openliberty.wsoc.2.1.internal_fat

**Comment by volosied** [comment]
#libby

**Pull request merged**